File: modules/android/examples/GetStartedJava/app/src/main/java/com/couchbase/android/getstarted/java/DBManager.java
//
// Copyright (c) 2023 Couchbase, Inc All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
package com.couchbase.android.getstarted.java;

import android.util.Log;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.concurrent.atomic.AtomicReference;

import com.couchbase.lite.BasicAuthenticator;
import com.couchbase.lite.Collection;
import com.couchbase.lite.CollectionConfiguration;
import com.couchbase.lite.CouchbaseLite;
import com.couchbase.lite.CouchbaseLiteException;
import com.couchbase.lite.DataSource;
import com.couchbase.lite.Database;
import com.couchbase.lite.Document;
import com.couchbase.lite.Expression;
import com.couchbase.lite.ListenerToken;
import com.couchbase.lite.MutableDocument;
import com.couchbase.lite.Query;
import com.couchbase.lite.QueryBuilder;
import com.couchbase.lite.Replicator;
import com.couchbase.lite.ReplicatorChangeListener;
import com.couchbase.lite.ReplicatorConfiguration;
import com.couchbase.lite.ReplicatorType;
import com.couchbase.lite.ResultSet;
import com.couchbase.lite.SelectResult;
import com.couchbase.lite.URLEndpoint;


@SuppressWarnings("unused")
public class DBManager {
    private static final String TAG = "START_JAVA";

    private static final AtomicReference<DBManager> INSTANCE = new AtomicReference<>();

    public static synchronized DBManager getInstance() {
        DBManager mgr = INSTANCE.get();
        if (mgr == null) {
            mgr = new DBManager();
            if (INSTANCE.compareAndSet(null, mgr)) { mgr.init(); }
        }
        return INSTANCE.get();
    }

    private Database database;
    private Collection collection;
    private Replicator replicator;

    private DBManager() { }

    // tag::getting-started[]

    // <.>
    // One-off initialization
    private void init() {
        CouchbaseLite.init(GettingStartedApplication.getAppContext());
        Log.i(TAG, "CBL Initialized");
    }

    // <.>
    // Create a database
    public void createDb(String dbName) throws CouchbaseLiteException {
        database = new Database(dbName);
        Log.i(TAG, "Database created: " + dbName);
    }

    // <.>
    // Create a new named collection (like a SQL table)
    // in the database's default scope.
    public void createCollection(String collName) throws CouchbaseLiteException {
        collection = database.createCollection(collName);
        Log.i(TAG, "Collection created: " + collection);
    }

    // <.>
    // Create a new document (i.e. a record)
    // and save it in a collection in the database.
    public String createDoc() throws CouchbaseLiteException {
        MutableDocument mutableDocument = new MutableDocument()
            .setFloat("version", 2.0f)
            .setString("language", "Java");
        collection.save(mutableDocument);
        return mutableDocument.getId();
    }

    // <.>
    // Retrieve immutable document and log the database generated
    // document ID and some document properties
    public void retrieveDoc(String docId) throws CouchbaseLiteException {
        Document document = collection.getDocument(docId);
        if (document == null) {
            Log.i(TAG, "No such document :: " + docId);
        }
        else {
            Log.i(TAG, "Document ID :: " + document.getId());
            Log.i(TAG, "Learning :: " + document.getString("language"));
        }
    }

    // <.>
    // Retrieve and update a document.
    public void updateDoc(String docId) throws CouchbaseLiteException {
        Document document = collection.getDocument(docId);
        if (document != null) {
            collection.save(
                document.toMutable().setString("language", "Kotlin"));
        }
    }

    // <.>
    // Create a query to fetch documents with language == Kotlin.
    public void queryDocs() throws CouchbaseLiteException {
        Query query = QueryBuilder.select(SelectResult.all())
            .from(DataSource.collection(collection))
            .where(Expression.property("language").equalTo(Expression.string("Kotlin")));

        try (ResultSet rs = query.execute()) {
            Log.i(TAG, "Number of rows :: " + rs.allResults().size());
        }
    }

    // <.>
    // OPTIONAL -- if you have Sync Gateway Installed you can try replication too.
    // Create a replicator to push and pull changes to and from the cloud.
    // Be sure to hold a reference somewhere to prevent the Replicator from being GCed
    public ListenerToken replicate(String uri, ReplicatorChangeListener listener) throws URISyntaxException {
        CollectionConfiguration collConfig = new CollectionConfiguration()
            .setPullFilter((doc, flags) -> "Java".equals(doc.getString("language")));

        ReplicatorConfiguration replConfig =
            new ReplicatorConfiguration(
                new URLEndpoint(new URI(uri)))
                .addCollection(collection, collConfig)
                .setType(ReplicatorType.PUSH_AND_PULL)
                .setAuthenticator(new BasicAuthenticator("sync-gateway", "password".toCharArray()));

        replicator = new Replicator(replConfig);

        // Listen to replicator change events.
        // Use `token.remove()` to stop the listener

        ListenerToken token = replicator.addChangeListener(listener);

        // Start replication.
        replicator.start();

        return token;
    }
    // end::getting-started[]
}

File: modules/android/examples/GetStartedKotlin/app/src/main/java/com/couchbase/android/getstarted/kotlin/DBManager.kt
//
// Copyright (c) 2023 Couchbase, Inc All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
package com.couchbase.android.getstarted.kotlin

import android.content.Context
import android.util.Log
import com.couchbase.lite.BasicAuthenticator
import com.couchbase.lite.Collection
import com.couchbase.lite.CollectionConfiguration
import com.couchbase.lite.CouchbaseLite
import com.couchbase.lite.DataSource
import com.couchbase.lite.Database
import com.couchbase.lite.Expression
import com.couchbase.lite.MutableDocument
import com.couchbase.lite.Query
import com.couchbase.lite.QueryBuilder
import com.couchbase.lite.Replicator
import com.couchbase.lite.ReplicatorChange
import com.couchbase.lite.ReplicatorConfigurationFactory
import com.couchbase.lite.ReplicatorType
import com.couchbase.lite.SelectResult
import com.couchbase.lite.URLEndpoint
import com.couchbase.lite.newConfig
import com.couchbase.lite.replicatorChangesFlow
import kotlinx.coroutines.flow.Flow
import java.net.URI
import java.util.concurrent.atomic.AtomicReference

class DBManager {
    private var database: Database? = null
    private var collection: Collection? = null
    private var replicator: Replicator? = null

    // tag::getting-started[]

    // <.>
    // One-off initialization
    private fun init(context: Context) {
        CouchbaseLite.init(context)
        Log.i(TAG, "CBL Initialized")

    }

    // <.>
    // Create a database
    fun createDb(dbName: String) {
        database = Database(dbName)
        Log.i(TAG, "Database created: $dbName")
    }

    // <.>
    // Create a new named collection (like a SQL table)
    // in the database's default scope.
    fun createCollection(collName: String) {
        collection = database!!.createCollection(collName)
        Log.i(TAG, "Collection created: $collection")
    }

    // <.>
    // Create a new document (i.e. a record)
    // and save it in a collection in the database.
    fun createDoc(): String {
        val mutableDocument = MutableDocument()
            .setFloat("version", 2.0f)
            .setString("language", "Java")
        collection?.save(mutableDocument)
        return mutableDocument.id
    }

    // <.>
    // Retrieve immutable document and log the database generated
    // document ID and some document properties
    fun retrieveDoc(docId: String) {
        collection?.getDocument(docId)
            ?.let {
                Log.i(TAG, "Document ID :: ${docId}")
                Log.i(TAG, "Learning :: ${it.getString("language")}")
            }
            ?: Log.i(TAG, "No such document :: $docId")
    }

    // <.>
    // Retrieve immutable document update `language` property
    // document ID and some document properties
    fun updateDoc(docId: String) {
        collection?.getDocument(docId)?.let {
            collection?.save(
                it.toMutable().setString("language", "Kotlin")
            )
        }
    }

    // <.>
    // Create a query to fetch documents with language == Kotlin.
    fun queryDocs() {
        val coll = collection ?: return
        val query: Query = QueryBuilder.select(SelectResult.all())
            .from(DataSource.collection(coll))
            .where(Expression.property("language").equalTo(Expression.string("Kotlin")))
        query.execute().use { rs ->
            Log.i(TAG, "Number of rows :: ${rs.allResults().size}")
        }
    }

    // <.>
    // OPTIONAL -- if you have Sync Gateway Installed you can try replication too.
    // Create a replicator to push and pull changes to and from the cloud.
    // Be sure to hold a reference somewhere to prevent the Replicator from being GCed
    fun replicate(): Flow<ReplicatorChange>? {
        val coll = collection ?: return null

        val collConfig = CollectionConfiguration()
            .setPullFilter { doc, _ -> "Java" == doc.getString("language") }

        val repl = Replicator(
            ReplicatorConfigurationFactory.newConfig(
                target = URLEndpoint(URI("ws://localhost:4984/getting-started-db")),
                collections = mapOf(setOf(coll) to collConfig),
                type = ReplicatorType.PUSH_AND_PULL,
                authenticator = BasicAuthenticator("sync-gateway", "password".toCharArray())
            )
        )

        // Listen to replicator change events.
        val changes = repl.replicatorChangesFlow()

        // Start replication.
        repl.start()
        replicator = repl

        return changes
    }
    // end::getting-started[]

    companion object {
        private const val TAG = "START_KOTLIN"

        private val INSTANCE = AtomicReference<DBManager?>()

        @Synchronized
        fun getInstance(context: Context): DBManager {
            var mgr = INSTANCE.get()
            if (mgr == null) {
                mgr = DBManager()
                if (INSTANCE.compareAndSet(null, mgr)) {
                    mgr.init(context)
                }
            }
            return INSTANCE.get()!!
        }
    }
}
